blog/2024-08-08-quorum-queues-in-4.0/index.md

```diff
@@ -0,0 +1,197 @@
+handle them. Classic queues require an argument (`x-max-priority`) which defines the maximum
+number of priorities for a given queue. Values from 0 to 255 are technically allowed, although
+no more than a handful of priorities per queue should really be used. Quorum queues don't require
+any upfront declaration (no need to enable priorities for a given queue), but there are exactly
+two priorities per queue: low/normal and high. The behavior matches that of AMQP 1.0 specification
+(see [chapter 3.2.1 of the AMQP 1.0 specification](https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-header)):
+* a priority value between 0 and 4 (inclusive) is treated as the low/normal priority
+* any value above 4 is considered a high priority
+* if the publisher doesn't specify the priority of a message, the value of `4` is assumed (low/normal priority)
+
+If a quorum queue contains both low and high priority messages, consumers will receive a mix of
+both, with a ratio of 2 high priority messages for every 1 low priority message. This approach
+avoids [starvation](https://en.wikipedia.org/wiki/Starvation_(computer_science)), since regardless
+of the number of high priority messages, a progress is made on processing the low priority messages as well.
+This is in contrast with the classic queue implementation, which will always deliver higher priority
+messages first, if there are any, and therefore the low priority messages may never get delivered (or, more likely,
+their delivery latency will be very high).
+
+Here's a visual representation of how this works. In preparation for this test, we first published 100k
+low priority messages and then 100k high priority messages. Since quorum queues were not priority-aware until 4.0,
+if we did that in an older version and then start a consumer, it would simply receive the low priority messages first
+(as they are older) and then all the high priority messages. With 4.0, we can see that the consumer
+immediately starts receiving a mix of about 1000 low priority messages per second and twice as many high priority
+messages, for a total of ~3500 messages per second (the actual delivery rates are not important here,
+they depend on many factors; the 2:1 high/low priority ratio is what matters in the context of priorities).
+Once the queue delivers all the high priority messages, the consumer starts receiving ~3500 low priority
+messages per second - as many as it can handle in this test scenario. The blue dotted line (with the axis
+scale on the right) is the number of ready messages in the queue (total for both priorities) - we can see it
+starts at 200k and ultimately drops to zero.
+
+![Consumer receives a mix of high (yellow) and low priority (green) messages](message-priorities-low-first.png)
+
+It should be intuitive, but let's consider the opposite scenario - what if we publish all the high priority
```

Comment:
I'd drop the "it should be intuitive" part. It may or may not be but here is how QQ priorities work in another example.